### DISCUSSION TITLE
Composite

### DISCUSSION THREAD

**Discussion by leanmendoza** [decentraland/sdk#603] [Ideas]
# Composite

## Intro and definition
A composite is a portion of a scene that only has component data (associated with entities). The scene has although code and with it, systems.

Definitions:
1. A composite has to have in its root entity the component **`CompositeRoot`**, this component has the next properties:
    - `name: string`: identify the composite and its value is used as the display name
    - `instanced?: boolean`: usually true, if false is because something went wrong (e.g. recursive instantiation)  
    - `mappedEntities?: Map<Entity, Entity>`: only used when it's instanced in SYNC mode, KEY is the entity in the composite, and the value is the target entity in the scene where is instanced
    - `override?: boolean`: if true, this composite can not be SYNC anymore
2. There can't be circular instantiation, e.g. CompositeA contains a CompositeB, I can not instance CompositeA in B nor itself. 
3. The composite is serialized by the CRDT protocol, all entities' versions must be 0, so entities' ids from 512 to 65535 are available.
4. The state of a composite with more composites inside that already has all component data, doesn't necessarily instance recursively.
5. In the instantiation process, the entity references must be updated with their mapped entities. It's necessary to identify what components (built-in and custom ones) need this.

## Example code to instance a composite
```ts
/**
 * Instance a composite and returns its root entity
 * @param compositeData state serialized by the CRDT protocol
 * @param getNextAvailableEntity function that gives unused entities
 * @param rootEntity (optional) suggested mapped rootEntity for the composite
 */
function instanceComposite(
    compositeData: Uint8Array, 
    getNextAvailableEntity: () => Entity | null, 
    rootEntity?: Entity
){
    const mappedEntities: Record<Entity, Entity> = {}

    /**
     * Return the entity mapping or fail if there is no more 
     */
    function getEntityMapping(compositeEntity: Entity): Entity {
        if (mappedEntities[compositeEntity]) {
            return mappedEntities[compositeEntity]
        }

        // This function in runtime can be just `engine.addEntity()`
        const entity = getNextAvailableEntity()
        if (entity === null) {
            throw new Error('There is no more entities to allocate')
        }
        mappedEntities[compositeEntity] = entity
        return entity
    }

    // Map rootEntity if exists
    if (rootEntity) {
        mappedEntities[0] = rootEntity 
    }

    // copy engine definition with empty state, this only replicates the component definitions
    const tempEngine = engine.instance()

    // this processes the Uint8Array and put all the data in the engine
    tempEngine.setState(compositeData)

    // Check if it's a composite
    const compositeCompositeRootComponent = tempEngine.getComponent(CompositeRoot.id)
    if (compositeCompositeRootComponent.getOrNull(tempEngine.RootEntity) === null) {
        throw new Error('The composite to instance must have a CompositeRoot at RootEntity')
    }

    // Copy all component data
    for (const componentDefinition of tempEngine.compIterator()) {
        for (const [entity, value] of tempEngine.getEntitiesWith(componentDefinition)) {
            const targetEntity = getEntityMapping(entity)
            engine.getComponent(componentDefinition.id).create(targetEntity, value)
        }
    }  

    // Check for entity reference in the components
    for (const componentDefinition of tempEngine.compIterator()) {
        for (const [entity, value] of tempEngine.getEntitiesWith(componentDefinition)) {
            const targetEntity = getEntityMapping(entity)
            if (Transform.id === componentDefinition.id) {
                if (value.parent) {
                    Transform.getMutable(targetEntity).parent = mappedEntities[value.parent] || null
                }
            } else {
                // TODO: with static reflection, look for `Schema.Entity` in custom components
            }
        }
    }

    const composite = CompositeRoot.getMutable(compositeRootEntity)

    // This would enable to receive changes, if changes are not necessary this can be ignored
    composite.mappedEntities = mappedEntities
    composite.instanced = true

    // With override true, the changes are no more propagated
    composite.override = false

    return compositeRootEntity
}

```

## Example at data level
# Scenario 1
- composite A:
```
root (e=0):
- CompositeRoot { name = "FunCubeComposite" }

e=513
- Transform { pos = (8, 4, 2), parent = null }

e=514
- EntityName { name = "My Cube"}
- MeshRenderer { cube }
- Transform { pos = (0, 0, 0), parent = 513 }
- Material { color = Red }

```

- composite B:
```
root (e=0):
- CompositeRoot { name = "ClickAudioComposite" }

e=513
- PointerEvent { name = "My Cube"}
- CustomAudioReference { audioEntity = 515 }

e=514
- AudioSource { src = "funny-click.wav" }

e=515
- AudioSource { src = "scarry-click.wav" }

```

Instancing these two composites, first composite-A and then composite-B the resulting state would be:

```
root (e=512):
- CompositeRoot { name = "ClickAudioComposite" }

e=513
- PointerEvent { name = "My Cube"}
- CustomAudioReference { audioEntity = 515 }

e=514
- AudioSource { src = "funny-click.wav" }

e=515
- AudioSource { src = "scarry-click.wav" }

root (e=516):
- CompositeRoot { name = "FunCubeComposite" }

e=517
- Transform { pos = (8, 4, 2), parent = null }

e=518
- EntityName { name = "My Cube"}
- MeshRenderer { cube }
- Transform { pos = (0, 0, 0), parent = 517 }
- Material { color = Red }

```

## Pending specs
1. Update a composite by name:
    - Remove all old data
    - Append new data and update the ones that stay but were modified
2. Fetch a composite by path, instance from the file

**Comment by menduz** [comment]
Awesome job, I'd like to bring up some points about instancing an extra engine. The most important one being the ergonomics and the statefullness of the components. Since components are instances *for a specific engine, you won't be able to access the original definitions when processing the messages.

Also, parsing the CRDT messages into the component definition and keeping the CRDT states and timestamp would be doing double work for literally zero gain. 

After careful thought, I think a simpler reader for entities may be better fit, with the benefit of also work with the editor as "standard composite format". It could have two representations: text & binary. And it would enable possible extensions that would be very complicated in the raw CRDT messages

**Comment by leanmendoza** [comment]
Modification from the original:

# Composite

## Intro and definition
A composite is a custom asset supported by the SDK7, at the first iteration it contains scene data that only has component data (associated with entities). 

Definitions:
1. A composite has to have in its root entity the component **`CompositeRoot`**, this component has the next properties:
    - `id: string`: identifier of the composite 
    - `instanced?: boolean`: usually true, if false is because something went wrong (e.g. recursive instantiation)  
    - `mappedEntities?: Map<Entity, Entity>`: only used when it's instanced in SYNC mode, KEY is the entity in the composite, and the value is the target entity in the scene where is instanced
    - `override?: boolean`: if true, this composite can not be SYNC anymore
2. There can't be circular instantiation, e.g. CompositeA contains a CompositeB, I can not instance CompositeA in B nor itself. 
3. The state of a composite with more composites inside that already has all component data, doesn't necessarily instance recursively.
4. In the instantiation process, the entity references must be updated with their mapped entities. It's necessary to identify what components (built-in and custom ones) need this.
5. There are two ways of serializing a composite, a human-readable format with JSON, and the binary that it'd be faster for deployments (and runtime loading). The binary one would be serialized by a protocol buffer.

## Composite example
```ts
type Composite = {
  id: string
  displayName?: string
  sceneData: {
    compoments: Record<
      string,
      {
        schemaDescription?: string
        data: Record<Entity, any>
      }
    >
  }
  externalDependencies?: string[]
}

const compositeA: Composite = {
  id: 'decentraland::example::cube-spawner',
  displayName: 'Decentraland Cube Spawner Composite',
  sceneData: {
    compoments: {
      'composite::root': {
        data: {
          0: { id: 'decentraland::example::cube-spawner' }
        }
      },
      'core::Transform': {
        data: {
          512: { position: { x: 8, y: 1, z: 8 } }
        }
      },
      'core::MeshRenderer': {
        data: {
          512: { mesh: { $case: 'box', box: {} } }
        }
      },
      'core::PointerEvents': {
        data: {
          512: { pointerEvents: { eventType: 1, eventInfo: { hoverText: 'Click to spawn a cube' } } }
        }
      }
    }
  }
}


```

**Reply by cazala** [reply]
Hello @leanmendoza, after reading this there are a few questions that come to my mind: 
- Do we need the `instanced` boolean, if we already know there are `mappedEntities`?
- Going one step further, do we need `mappedEntities` if we just reference the instaced composite, and we have access to all the composites from the renderer?
- Do we need the composite to have a `composite::root` component with the composite id on its root entity? Can't we just reference it from the scene that is instanciating it by the id of the composite itself?
- Do we need `sceneData` if we only have `components` in there? Could't we unwrap it?

After asking myself these questions, I came back with a proposal that I think simplifies this a bit, and it takes into consideration some things that I believe are being left out, like the ability to distribute composites and the ability to override only some components while keeping others in sync.

Lets start with the definition of a `Package`, which is the minimal unit that the Inspector can work with (the local workspace would be a `Package`, then others could be imported from `node_modules` or a remote server like the marketplace).

```ts 
type Package = {
  composites: Composite[],
  assets: Asset[]
}
```

Then we have composites, which are a bag of components data for each entity + their schema

```ts
type Composite = {
  id: string;
  name: string; // meant for display
  components: Component[];
};

// A component contains data its schema definition
type Component = {
  id: string
  schema: any; // this is the definition of the schema, somehting like JSON Schema
  data: Record<number, any>; // here are the actual values for each entity
};
```

Note: I made a change here converting the component map into an array. This was because I think it is important to have the `id` stored within the `Component` itself, otherwise this piece of data is living outside (ie just by looking at a single component you can't tell which is it, you need the parent map of components, this is cumbersome when it comes to rendering a single component). We could still keep it as a map instead of an array and have the id duplicated (within the component shape, and outside as the key in the map). I have no strong opinions about one or the other, I just went for the one without duplication.

This is an example of how the local scene in the workspace would look, with just a cube in the middle:

```ts
const local: Package = {
  composites: [
    {
      id: "scene",
      name: "My Cube Scene",
      components: [
        {
          id: "core::Transform",
          schema: null,
          data: {
            512: {
              position: {
                x: 8,
                y: 1,
                z: 8,
              },
            },
          },
        },
        { 
          id: "core::MeshRenderer",
          schema: null,
          data: {
            512: { mesh: { $case: "box", box: {} } },
          },
        },
      ],
    },
  ],
  assets: [],
};
```

Now this is how it would look, if we move the cube two its own composite, and we instanciate it twice in the scene, and scaled one of the instances 2x:

```ts
const local: Package = {
  composites: [
    {
      id: "scene",
      name: "My Scene With Cubes",
      components: [
        {
          id: "composite::root",
          schema: null,
          data: {
            512: { instanceOf: "my-cube", overrides: [] },
            513: { instanceOf: "my-cube", overrides: [] },
          },
        },
        // Here we scale the second instance
        {
          id: "core::Transform",
          schema: null,
          data: {
            513: {
              scale: {
                x: 2,
                y: 2,
                z: 2,
              },
            },
          },
        },
      ],
    },
    {
      id: "my-cube",
      name: "My Cube Composite",
      components: [
        {
          id: "core::Transform",
          schema: null,
          data: {
            512: {
              position: {
                x: 8,
                y: 1,
                z: 8,
              },
            },
          },
        },
        {
          id: "core::MeshRenderer",
          schema: null,
          data: {
            512: { mesh: { $case: "box", box: {} } },
          },
        },
      ],
    },
  ],
  assets: [],
};
```

The shape of the data stored in the `composite::root` component is as follows:

```ts
type CompositeRootComponent = {
  instanceOf: string; // id pointing to the composite
  overrides: Component[] // this is the same shape as the composite.components map, here we can store overrides for particular components for specific entities
};
```
This is how we override the instance that was scaled, to make the cube be a sphere

```ts
const local: Package = {
  composites: [
    {
      id: "scene",
      name: "My Scene With Cubes",
      components: [
        {
          id: "composite::root",
          schema: null,
          data: {
            512: { instanceOf: "my-cube", overrides: [] },
            513: { instanceOf: "my-cube", overrides: [
              {
                id: "core::MeshRenderer",
                schema: null,
                data: {
                  // The id of this entity points to the one in the composite, not the one in the base scene. It would end up being entity 513 after mapping it.
                  512: { mesh: { $case: "box", box: {} } },
                },
              },
            ] },
          },
        },
        {
          id: "core::Transform",
          schema: null,
          data: {
            513: {
              scale: {
                x: 2,
                y: 2,
                z: 2,
              },
            },
          },
        },
      ],
    },
    {
      id: "my-cube",
      name: "My Cube Composte",
      components: [
        {
          id: "core::Transform",
          schema: null,
          data: {
            512: {
              position: {
                x: 8,
                y: 1,
                z: 8,
              },
            },
          },
        },
        {
          id: "core::MeshRenderer",
          schema: null,
          data: {
            512: { mesh: { $case: "box", box: {} } },
          },
        },
      ],
    },
  ],
  assets: [],
};
```

One final note:

I left it out for simplicity, but I think the full type of a `Package` should also include a `version` and an `id`, for two reasons:
- The shape of the `Package` could change in the future, but if there are packages distributed already (via npm packages or marketplace NFTs or whatever) we need to be able to migrate them in runtime and support them, thus the `version`.
- Then we also need an `id` in case there are clashing id among the `composites` and `assets` of two different packages, so we can namespace them. So the full type of a project would be like:

```ts
type Package = {
  version: 1;
  id: string;
  composites: Composite[];
  assets: Asset[];
};
```

Then in the future we could:

```ts
type PackageV1 = {
  version: 1;
  id: string;
  composites: CompositeV1[];
  assets: AssetV1[];
};

type PackageV2 = {
  version: 2;
  id: string;
  composites: CompositeV2[];
  assets: AssetV2[];
};

type Project = ProjectV1 | ProjectV2
```
